File: src/aoxcon_ref/envelope.py
import hashlib
from dataclasses import dataclass
from typing import Dict, Any, Set, Tuple

from . import errors

REQUIRED_FIELDS = {
    "version",
    "message_id",
    "source_chain",
    "target_chain",
    "source_app",
    "target_app",
    "actor_id",
    "nonce",
    "timestamp",
    "not_before",
    "expiry",
    "payload_hash",
    "proof_type",
    "proof",
}


@dataclass
class InMemoryReplayStore:
    processed: Set[str]
    nonces: Dict[Tuple[str, str], int]

    def __init__(self) -> None:
        self.processed = set()
        self.nonces = {}


def _is_chain_id(value: str) -> bool:
    return isinstance(value, str) and ":" in value and len(value.split(":")) == 2


def _hash_payload(payload: bytes) -> str:
    return "0x" + hashlib.sha256(payload).hexdigest()


def process_envelope(
    envelope: Dict[str, Any],
    payload: bytes,
    now_ts: int,
    replay: InMemoryReplayStore,
    authorized_actors: Set[str],
) -> str:
    if not REQUIRED_FIELDS.issubset(set(envelope.keys())):
        return errors.ERR_INVALID_SCHEMA

    if not _is_chain_id(envelope["source_chain"]) or not _is_chain_id(envelope["target_chain"]):
        return errors.ERR_INVALID_CHAIN_ID

    message_id = envelope["message_id"]
    if message_id in replay.processed:
        return errors.OK_ALREADY_PROCESSED

    key = (envelope["source_chain"], envelope["actor_id"])
    last_nonce = replay.nonces.get(key, -1)
    if envelope["nonce"] <= last_nonce:
        return errors.ERR_INVALID_NONCE

    if now_ts < envelope["not_before"]:
        return errors.ERR_STALE_MESSAGE
    if now_ts > envelope["expiry"]:
        return errors.ERR_EXPIRED_MESSAGE

    if envelope["payload_hash"] != _hash_payload(payload):
        return errors.ERR_PROCESSING_FAILED

    if envelope["proof_type"] not in {"merkle", "sigset", "lightclient"} or not envelope["proof"]:
        return errors.ERR_INVALID_PROOF

    if envelope["actor_id"] not in authorized_actors:
        return errors.ERR_UNAUTHORIZED_ACTOR

    replay.processed.add(message_id)
    replay.nonces[key] = envelope["nonce"]
    return errors.OK_PROCESSED

File: haskell/aoxcon-cardano-ref/src/Aoxcon/Envelope.hs
module Aoxcon.Envelope where

import Aoxcon.Errors
import Aoxcon.Types
import Crypto.Hash (Digest, SHA256, hash)
import qualified Data.ByteArray.Encoding as BA
import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BS8
import qualified Data.Map.Strict as Map
import qualified Data.Set as Set
import qualified Data.Text as T

isChainId :: String -> Bool
isChainId value =
  case break (== ':') value of
    ("", _) -> False
    (_, "") -> False
    (_, ':':rest) -> not (null rest) && all (/= ':') rest
    _ -> False

hashPayload :: BS.ByteString -> String
hashPayload payload =
  let digest :: Digest SHA256
      digest = hash payload
   in "0x" <> BS8.unpack (BA.convertToBase BA.Base16 digest)

validProofType :: String -> Bool
validProofType p = p `elem` ["merkle", "sigset", "lightclient"]

processEnvelope
  :: Envelope
  -> BS.ByteString
  -> Int
  -> ReplayStore
  -> Set.Set String
  -> (String, ReplayStore)
processEnvelope env payload nowTs replay authorizedActors
  | not (isChainId $ sourceChain env) || not (isChainId $ targetChain env) = (errInvalidChainId, replay)
  | Set.member (messageId env) (processed replay) = (okAlreadyProcessed, replay)
  | nonce env <= lastNonce = (errInvalidNonce, replay)
  | nowTs < notBefore env = (errStaleMessage, replay)
  | nowTs > expiry env = (errExpiredMessage, replay)
  | payloadHash env /= hashPayload payload = (errProcessingFailed, replay)
  | not (validProofType $ proofType env) || null (proof env) = (errInvalidProof, replay)
  | not (Set.member (actorId env) authorizedActors) = (errUnauthorizedActor, replay)
  | otherwise =
      let newReplay =
            ReplayStore
              { processed = Set.insert (messageId env) (processed replay)
              , nonces = Map.insert key (nonce env) (nonces replay)
              }
       in (okProcessed, newReplay)
  where
    key = (sourceChain env, actorId env)
    lastNonce = Map.findWithDefault (-1) key (nonces replay)

mkEnvelope
  :: String -> String -> String -> String -> String -> String -> String
  -> Int -> Int -> Int -> Int -> String -> String -> String
  -> Envelope
mkEnvelope ver mid srcChain trgChain srcApp trgApp actor nonceVal ts nb expTs pHash pType p =
  Envelope
    { version = ver
    , messageId = mid
    , sourceChain = srcChain
    , targetChain = trgChain
    , sourceApp = srcApp
    , targetApp = trgApp
    , actorId = actor
    , nonce = nonceVal
    , timestamp = ts
    , notBefore = nb
    , expiry = expTs
    , payloadHash = pHash
    , proofType = pType
    , proof = p
    }

mkPayloadHash :: T.Text -> String
mkPayloadHash = hashPayload . BS8.pack . T.unpack
